rmgpy/data/kinetics/family.py

```diff
@@ -3068,68 +3095,275 @@ def generateTree(self, obj=None, thermoDatabase=None, T=1000.0):
+
+
+    def makeTreeNodes(self,templateRxnMap=None,obj=None,T=1000.0,nprocs=0,depth=0,minSplitableEntryNum=2,minRxnsToSpawn=20):
+
+        if depth > 0:
+            root = self.groups.entries[templateRxnMap.keys()[0]]
+        else:
+            for entry in self.groups.entries.values(): #find the root entry for this branch
+                if entry.index != -1:
+                    root = entry
+                    break
+            while root.parent is not None:
+                root = root.parent
+                
+        psize = float(len(templateRxnMap[root.label]))
         
         multCompletedNodes = [] #nodes containing multiple identical training reactions
         boo = True #if the for loop doesn't break becomes false and the while loop terminates
+        activeProcs = []
+        activeConns = []
+        activeProcNum = []
+        procNames = []
+        freeProcs = nprocs
+        extraEntries = []
+
         while boo:
-            for entry in self.groups.entries.itervalues():
+            removeInds = []
+            for k,p in enumerate(activeProcs): #check if any processes have finished
+                if activeConns[k].poll():
+                    newEntries = self.absorbProcess(p,activeConns[k],procNames[k])
+                    extraEntries += newEntries
+                    removeInds.append(k)
+
+            removeInds.reverse()
+            for ind in removeInds: #remove finished process objects
+                freeProcs += activeProcNum[ind]
+                del activeProcNum[ind]
+                del activeProcs[ind]
+                del activeConns[ind]
+                del procNames[ind]
+
+            splitableEntryNum = 0
+            for label,items in templateRxnMap.iteritems(): #figure out how many splitable objects there are
+                entry = self.groups.entries[label]
+                if len(items) > 1 and entry not in multCompletedNodes:
+                    splitableEntryNum += 1
+
+            for label in templateRxnMap.keys():
+                entry = self.groups.entries[label]
                 if not isinstance(entry.item, Group): #skip logic nodes
                     continue
+                if psize == 0.0:
+                    continue
                 if entry.index != -1 and len(templateRxnMap[entry.label])>1 and entry not in multCompletedNodes:
-                    boo2 = self.extendNode(entry,templateRxnMap,thermoDatabase,obj,T)
+                    if freeProcs > 0 and splitableEntryNum > minSplitableEntryNum and len(templateRxnMap[entry.label])>minRxnsToSpawn:
+                        procsOut = int(len(templateRxnMap[entry.label])/psize*freeProcs)
+                        freeProcs -= procsOut
+                        assert freeProcs >= 0
+                        conn,p,name = spawnTreeProcess(family=self,templateRxnMap={entry.label:templateRxnMap[entry.label]},obj=obj,T=T,nprocs=procsOut-1,depth=depth,minSplitableEntryNum=minSplitableEntryNum,minRxnsToSpawn=minRxnsToSpawn)
```

Comment:
Could you add a line break (or two) in here? (And spaces after commas, if you don't mind.)

Reply:
done